library/src/contraction/contraction_solution.cpp

```diff
@@ -24,10 +24,140 @@
  *
  *******************************************************************************/
 
+#include <set>
+
 #include "contraction_solution.hpp"
+#include "util.hpp"
 
 namespace hiptensor
 {
+    std::array<std::vector<std::size_t>, 8>
+        normalizeTensorModes(std::vector<std::size_t> const& a_ms_ks_lengths,
+                             std::vector<std::size_t> const& a_ms_ks_strides,
+                             std::vector<int32_t> const&     a_ms_ks_modes,
+                             std::vector<std::size_t> const& b_ns_ks_lengths,
+                             std::vector<std::size_t> const& b_ns_ks_strides,
+                             std::vector<int32_t> const&     b_ns_ks_modes,
+                             std::vector<std::size_t> const& e_ms_ns_lengths,
+                             std::vector<std::size_t> const& e_ms_ns_strides,
+                             std::vector<int32_t> const&     e_ms_ns_modes)
+    {
+        std::vector<std::size_t> normal_a_ms_ks_lengths(MaxNumDimsM + MaxNumDimsK, 1);
+        std::vector<std::size_t> normal_a_ms_ks_strides(MaxNumDimsM + MaxNumDimsK, 1);
+        std::vector<int32_t>     normal_a_ms_ks_modes(MaxNumDimsM + MaxNumDimsK, -1);
+        std::vector<std::size_t> normal_b_ns_ks_lengths(MaxNumDimsK + MaxNumDimsN, 1);
+        std::vector<std::size_t> normal_b_ns_ks_strides(MaxNumDimsK + MaxNumDimsN, 1);
+        std::vector<int32_t>     normal_b_ns_ks_modes(MaxNumDimsK + MaxNumDimsN, -1);
+        std::vector<std::size_t> normal_e_ms_ns_lengths(MaxNumDimsM + MaxNumDimsN, 1);
+        std::vector<std::size_t> normal_e_ms_ns_strides(MaxNumDimsM + MaxNumDimsN, 1);
+        std::vector<int32_t>     normal_e_ms_ns_modes(MaxNumDimsM + MaxNumDimsN, -1);
+        int                      mOffset = 0;
+        int                      nOffset = 0;
+
+        // reorder m, n in A, B
+        for(int i = 0; i < e_ms_ns_modes.size(); i++)
```

Comment:
Is it possible to explain each of these steps a bit more? And possibly with a sample of the case they cover?

Reply:
![image](https://github.com/ROCm/hipTensor/assets/142121551/8bdbe328-0c6c-40a7-8098-392ac51ba56e)

Example:
|Matrix | mode | length|
|-|-|-|
|A | [3, 0, 4, 1] | [ 2, 3, 4, 5] |
|B | [4, 2, 3] | [ 4, 6, 2] |
|E | [2, 1, 0] | [6, 5, 3] |

**find out M,N,K**  
    mode_E ∩ mode_A == M
    mode_E ∩ mode_B == N
    mode_A ∩ mode_B == K

And we hold the order of M and N in E, order of K in A


|Matrix | mode | 
|-|-|
|A | [K0, M1, K1, M0]
|B | [K1, N0, K0] | 
|E | [N0, M0, M1] | 

**We need to permute and extend the matrices to have**

|Matrix | mode | 
|-|-|
|normalized A | [M0, M1, M2, M3, M4, M5, K0, K1, K2, K3, K4, K5] |
|normalized  B | [N0, N1, N2, N3, N4, N5, K0, K1, K2, K3, K4, K5] | 

**Line 60 - 79 change `normalized A` and `normalized B` to** 
|Matrix | mode | 
|-|-|
|normalized A | [M0, M1] |
|normalized  B | [N0] | 

**Line 81 - 91 extend `normalized A` and `normalized B` to** 

extended dimension: length is 1 and stride is the same as M1 and N0's stride.
|Matrix | mode | 
|-|-|
|normalized A | [M0, M1, M2, M3, M4, M5] |
|normalized  B | [N0, N1, N2, N3, N4, N5] | 

**Line 94 - 110 add K  to `normalized A` and `normalized B` ** 
|Matrix | mode | 
|-|-|
|normalized A | [M0, M1, M2, M3, M4, M5, K0, K1] |
|normalized  B | [N0, N1, N2, N3, N4, N5, K0, K1] | 

**Line 112 - 122 extend K in `normalized A` and `normalized B` ** 
|Matrix | mode | 
|-|-|
|normalized A | [M0, M1, M2, M3, M4, M5, K0, K1, K2, K3, K4, K5] |
|normalized  B | [N0, N1, N2, N3, N4, N5, K0, K1, K2, K3, K4, K5] | 

**Line 125 - 133 generate `normalized E` ** 

Copy modes from `normalized A` and `normalized B`
|Matrix | mode | 
|-|-|
|E | [N0, M0, M1] | 
|normalized E | [M0, M1, M2, M3, M4, M5, N0, N1, N2, N3, N4, N5] |

**Line 133 - 148 permute the E by switching length and strides **